File: app/src/main/java/org/bookdash/android/presentation/readbook/BookDetailActivity.java
package org.bookdash.android.presentation.readbook;

import android.os.Bundle;
import android.support.v4.view.ViewPager;
import android.view.ViewTreeObserver;
import android.view.WindowManager;

import org.bookdash.android.R;
import org.bookdash.android.domain.model.gson.BookPages;
import org.bookdash.android.presentation.activity.BaseAppCompatActivity;


public class BookDetailActivity extends BaseAppCompatActivity {
    public static final String BOOK_PAGES = "book_pages";
    public static final String LOCATION_BOOK = "book_location";
    public static final String BOOK_ARG = "book_obj";
    private ViewPager viewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_book_details);
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        viewPager = findViewById(R.id.view_pager);
        viewPager.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
            @Override
            public boolean onPreDraw() {
                viewPager.getViewTreeObserver().removeOnPreDrawListener(this);
                viewPager.setTranslationX(viewPager.getMeasuredWidth());
                viewPager.animate().translationXBy(-viewPager.getMeasuredWidth()).setDuration(1000);
                return true;
            }
        });
        String book = getIntent().getStringExtra(BOOK_ARG);
        BookPages bookPages = getIntent().getParcelableExtra(BOOK_PAGES);
        bookPages.getPages().add(0, null);
        String bookLocation = getIntent().getStringExtra(LOCATION_BOOK);
        PageAdapter pageAdapter = new PageAdapter(getSupportFragmentManager(), bookPages.getPages(),
                bookLocation);

        viewPager.setAdapter(pageAdapter);
        viewPager.setOffscreenPageLimit(1);


    }

    @Override
    protected String getScreenName() {
        return "BookDetailScreen";
    }


}

File: app/src/main/java/org/bookdash/android/presentation/readbook/BookDetailActivity.kt
package org.bookdash.android.presentation.readbook

import android.os.Bundle
import android.view.ViewTreeObserver
import android.view.WindowManager
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.viewpager.widget.ViewPager
import com.google.firebase.crashlytics.FirebaseCrashlytics
import org.bookdash.android.R
import org.bookdash.android.domain.model.gson.BookPages
import org.bookdash.android.presentation.activity.BaseAppCompatActivity


class BookDetailActivity : BaseAppCompatActivity() {
    private lateinit var viewPager: ViewPager

    private lateinit var pageAdapter: PageAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_book_details)
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        viewPager = findViewById(R.id.view_pager)
        viewPager.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
            override fun onPreDraw(): Boolean {
                viewPager.viewTreeObserver.removeOnPreDrawListener(this)
                viewPager.translationX = viewPager.measuredWidth.toFloat()
                viewPager.animate().translationXBy((-viewPager.measuredWidth).toFloat()).duration = 1000
                return true
            }
        })
        //val book = intent.getStringExtra(BOOK_ARG)
        val bookPages = intent.getParcelableExtra<BookPages>(BOOK_PAGES)
        if (bookPages == null) {
            onBookPagesNull()
            return
        }
        bookPages.pages.add(0, null)
        val bookLocation = intent.getStringExtra(LOCATION_BOOK)
        pageAdapter = PageAdapter(supportFragmentManager, bookPages.pages,
                bookLocation)

        viewPager.adapter = pageAdapter
        viewPager.offscreenPageLimit = 1

        setupViewModelListeners()
    }

    private fun setupViewModelListeners() {
        val readBookViewModel = ViewModelProviders.of(this).get<ReadBookViewModel>(ReadBookViewModel::class.java)
        readBookViewModel.pageBackEventTrigger.observe(this, Observer {
            if (viewPager.currentItem != 0) {
                viewPager.setCurrentItem(viewPager.currentItem - 1, true)
            }
        })
        readBookViewModel.pageForwardEventTrigger.observe(this, Observer {
            if (viewPager.currentItem != viewPager.childCount - 1) {
                viewPager.setCurrentItem(viewPager.currentItem + 1, true)
            }
        })
    }

    /**
     * Called when this Activity was launched with no book pages.
     * TODO Remove this method once bookPages == null bug is fixed.
     */
    private fun onBookPagesNull() {
        FirebaseCrashlytics.getInstance().recordException(Exception("Book pages null when opening book detail."))
        Toast.makeText(applicationContext, R.string.failed_to_open_book, Toast.LENGTH_LONG).show()
        finish()
    }

    override fun getScreenName(): String {
        return "BookDetailScreen"
    }

    companion object {
        val BOOK_PAGES = "book_pages"
        val LOCATION_BOOK = "book_location"
        val BOOK_ARG = "book_obj"
    }


}
